site/docs/plugins/commands.md

```diff
@@ -5,10 +5,628 @@ next: false
+- Setting custom behavior for commands that explicitly mention your bot's user,
+  like: `/start@your_bot`
+- Custom command prefixes, e.g: `+`, `?` or any symbol instead of `/`
+- Support for commands that are not in the beginning of the message
+- RegExp Commands!
+
+All of these features are made possible because you will define one or more central command structures that define your bot's commands.
+
+## Basic Usage
+
+Before we dive in, take a look at how you can register and handle a command with the plugin:
+
+```ts
+const myCommands = new CommandGroup();
+
+myCommands.command("hello", "Say hello", (ctx) => ctx.reply(`Hello, world!`));
+
+bot.use(myCommands);
+```
+
+This registers a new `/hello` command to your bot that will be handled by the given middleware.
+
+Now, let's get into some of the extra tools this plugin has to offer.
+
+## Importing
+
+First of all, here's how you can import all the necessary types and classes the plugin provides.
+
+::: code-group
+
+```ts [TypeScript]
+import {
+  CommandGroup,
+  commandNotFound,
+  commands,
+  type CommandsFlavor,
+} from "@grammyjs/commands";
+```
+
+```js [JavaScript]
+const { CommandGroup, commands, commandNotFound } = require(
+  "@grammyjs/commands",
+);
+```
+
+```ts [Deno]
+import {
+  CommandGroup,
+  commandNotFound,
+  commands,
+  type CommandsFlavor,
+} from "https://deno.land/x/grammy_commands/mod.ts";
+```
+
+:::
+
+Now that the imports are settled, let's see how we can make our commands visible to our users.
+
+## User Command Menu Setting
+
+Once you defined your commands with an instance of the `CommandGroup` class, you can call the `setCommands` method, which will register all the defined commands to your bot.
+
+```ts
+const myCommands = new CommandGroup();
+
+myCommands.command("hello", "Say hello", (ctx) => ctx.reply("Hi there!"));
+myCommands.command("start", "Start the bot", (ctx) => ctx.reply("Starting..."));
+
+bot.use(myCommands);
+
+await myCommands.setCommands(bot);
+```
+
+This will make it so every command you registered is displayed on the menu of a private chat with your bot, or whenever users type `/` on a chat your bot is a member of.
+
+### Context Shortcut
+
+What if you want some commands to be displayed only to certain users? For example, imagine you have a `login` and a `logout` command.
+The `login` command should only appear for logged out users, and vice versa.
+This is how you can do that with the commands plugin:
+
+::: code-group
+
+```ts [TypeScript]
+// Use the flavor to create a custom context
+type MyContext = Context & CommandsFlavor;
+
+// Use the new context to instantiate your bot
+const bot = new Bot<MyContext>("token");
+
+// Register the context shortcut
+bot.use(commands());
+
+const loggedOutCommands = new CommandGroup();
+const loggedInCommands = new CommandGroup();
+
+loggedOutCommands.command(
+  "login",
+  "Start your session with the bot",
+  async (ctx) => {
+    await ctx.setMyCommands(loggedInCommands);
```

Comment:
I received an error when running this code snippet.

Error:
```
Property 'setMyCommands' does not exist on type 'CommandContext<Context>'.
  Property 'setMyCommands' does not exist on type 'FilteredContext<NarrowMatch<Context, string>, Update & Record<"message", Record<"entities" | AddTwins<"message", "entities">, NotUndefined> & Partial<...>> & Partial<...>>'.
```
The code:
```ts
import { Bot, Context } from "https://deno.land/x/grammy@v1.31.0/mod.ts";
import {
  CommandGroup,
  commands,
  type CommandsFlavor,
} from "https://deno.land/x/grammy_commands@v0.10.0/mod.ts";

const botToken = Deno.env.get("BOT_TOKEN");
if (!botToken) throw new Error("Missing bot token!");

type MyContext = Context & CommandsFlavor;
const bot = new Bot<MyContext>(botToken);

bot.use(commands());

const loggedOutCommands = new CommandGroup();
const loggedInCommands = new CommandGroup();

loggedOutCommands.command(
  "login",
  "Start your session with the bot",
  async (ctx) => {
    await ctx.setMyCommands(loggedInCommands); // <-- ERROR
    await ctx.reply("Welcome! Session started!");
  },
);

// Rest of code...
```

Reply:
@roziscoding, the same error is still present in the latest version ([grammy_commands@0.11.0](https://deno.land/x/grammy_commands@0.11.0/mod.ts))

Reply:
Try
```ts
new CommandGroup<MyContext>();
```

Reply:
> Try
> 
> ```ts
> new CommandGroup<MyContext>();
> ```

It’s giving us another error:

```ts
const loggedOutCommands = new CommandGroup<MyContext>();
const loggedInCommands = new CommandGroup<MyContext>();

loggedOutCommands.command(
  "login",
  "Start your session with the bot",
  async (ctx) => {
    await ctx.setMyCommands(loggedInCommands);
    await ctx.reply("Welcome! Session started!");
  },
);
```
```
Argument of type 'CommandGroup<MyContext>' is not assignable to parameter of type 'CommandGroup<Context> | CommandGroup<Context>[]'.
  Type 'CommandGroup<MyContext>' is not assignable to type 'CommandGroup<Context>'.
    Types of property '_scopes' are incompatible.
      Type 'Map<string, Command<MyContext>[]>' is not assignable to type 'Map<string, Command<Context>[]>'.
        Type 'Command<MyContext>[]' is not assignable to type 'Command<Context>[]'.
          Type 'Command<MyContext>' is not assignable to type 'Command<Context>'.
            The types of '_composer.handler' are incompatible between these types.
              Type 'MiddlewareFn<MyContext>' is not assignable to type 'MiddlewareFn<Context>'.
                Type 'Context' is not assignable to type 'MyContext'.
                  Type 'Context' is missing the following properties from type 'CommandsFlavor<Context>': setMyCommands, getNearestCommand, getCommandEntities deno-ts(2345)
```

Reply:
Alright. I have no idea then. It's something that has to be fixed either in the code example or in the plugin.

Reply:
I'm running the snippet with no errors. Reload cache?

Reply:
should be 

```const loggedOutCommands = new CommandGroup<MyContext>()```

Reply:
That's what I also thought, but apparent that does not fix it, does it?